### ISSUE TITLE
Quality indicators for grammars

### ISSUE THREAD

**Issue by julianthome** [tree-sitter/tree-sitter#1094]
Except for the differentiation between _"fairly complete"_ and _"in development"_ grammars on the website (https://tree-sitter.github.io/tree-sitter/#available-parsers), I could not find any quality indicators for grammars. If there are already discussions around quality metrics for tree-sitter grammars, I would be grateful if you could point me to them. 

One possible quality indicator that could be helpful to understand the maturity of a given grammar could be language coverage (i.e., how many constructs of the target language are (un)-supported). 

This type of information would be very useful to identify, track and eliminate weak spots certain grammars may have.

**Comment by ahlinc** [comment]
I don't know about such discussions unfortunately but I'd like to describe my thoughts how this may work:
- Firstly, the simple list in documentation don't represent well any quality metrics.
- At this point amount of grammar implementations achieved a point where there're concurrent implementations for the same language exists.
- I think  several additional records may be introduced in the grammars `package.json` file like:
  - `specification_urls` - point to a language specification sources used for grammar implementation.
  - `completeness.percentage` - a percentage of a grammar completeness estimation provided by a grammar author.
  - `completeness.notes` - an array of notes about completeness provided by a grammar author.
  - `rule_naming_style` - `original_spec` - used naming based on a language spec, `tree_sitter_convention` - a snake  case style used in tree-sitter by convention. This may be useful to understand how much work needed to switch from a one grammar implementation to some another.

Such notes may be provided by grammar authors to track a grammar implementation state.
Most grammars hosted on Github can be fetched from the `tree-sitter` package [`dependants`](https://github.com/tree-sitter/tree-sitter/network/dependents?package_id=UGFja2FnZS0zMjE4MDA0NzQ1) and their `package.json` can be parsed for above mentioned fields automatically.
Also automatically may be computed some metrics like:
- Peak memory usage on grammar generation by `tree-sitter generate` command.
- Grammar binary library size.
- Does it have an external scanner.
- How much terminals handled by an external scanner.
- What fields about grammar filled in `package.json` like `scope`, `file_types` and so on.

All above information may be a source for automatically generated table of grammars instead of manually managed list.

> One possible quality indicator that could be helpful to understand the maturity of a given grammar could be language coverage (i.e., how many constructs of the target language are (un)-supported).

Such honest counting isn't an easy task because language specifications may be totally different forms and not all specs have parsable definitions like `BNF` notation. But in cases where it's possible it would be awesome to have a computed completeness estimation.

_In any case it would be great to have a discussion about a list of useful quality metrics and how they may be achieved._

/cc: @maxbrunsfeld 
/cc: @dcreager

**Comment by mattmassicotte** [comment]
I'd really like to see further progress here. I've used ~ 20 parsers myself. The worse issues I've run into are abandoned projects (like Markdown) and difficulty integrating parsers into my project. To address the latter, I've done a fair bit of work adding Makefiles and/or Swift package manager support.

It appears that many projects internalize parser building and query definitions instead of attempting to centralize them back into the parser projects themselves. Unsure how to address that one.

I think a pretty useful starting point would just be having one single, authoritative list. No commentary needed, just a big long list so people can at least find parsers more easily.

**Comment by ahelwer** [comment]
Who is responsible for curating the list? Every time I've seen one of these sorts of lists they're outdated within a year. Everyone wants to put their project on the list for the clout but many of them subsequently lose interest in maintaining it. Evaluating quality of a FOSS project is not a problem unique to tree-sitter.

**Comment by mattmassicotte** [comment]
Unfortunately, that list is suffering from exactly the problem you are talking about. It is out of date - the Markdown parser listed, for example, is abandoned. I considered updating it, like I did for the Swift entry, but was confused about the criteria for stable/development. Which is exactly what this issue seems to be about, and that's really cool!

I didn't mean to suggest something unwanted. I'm sure you all have a much better sense of what you want here. I only chimed in because @ahlinc mentioned it to me in #1894.

**Comment by ahelwer** [comment]
Well certainly if you would like to take ownership of curating the list it would be appreciated! Currently it is both incomplete and outdated.

**Comment by mattmassicotte** [comment]
Sorry, I just want to make sure I understand the suggestion: Are you saying I should update the list and also decide which parsers should be included/excluded?

If the project is looking for someone to do it, I would be happy to do this. However, I think I'd need some guidance on how to decide when/if a parser should be excluded. Swift was easy. The project under the tree-sitter org was marked as abandoned, really incomplete, and the replacement is very actively maintained. The two Markdown parsers I know of (https://github.com/MDeiml/tree-sitter-markdown and https://github.com/ikatyang/tree-sitter-markdown) are much less clear-cut. They both have very different designs, and I could see someone still wanting the original, even if it is no longer maintained.

**Comment by ahelwer** [comment]
Well all of us here are volunteers. Max is I guess the closest you can get to an "owner" of this repo but he is busy with the zed editor startup. So really you would decide all these things for yourself and as long as your decisions aren't wild enough for someone to make a big fuss they'd probably stand.

**Comment by clason** [comment]
As one of the maintainers for the nvim-treesitter (and Neovim core) project, I am quite interested in these questions myself, as nvim-treesitter is wrestling with the same issues.

Regarding centralization, I fully agree: I would love to see parsers handling their own build setup (via Makefile or CMake if necessary -- or something provided by tree-sitter itself (`tree-sitter build`)), as a small but non-negligible number of parsers require non-standard C(++) compiler options due to their external scanner.

I also would love to see projects developing (reference) queries together with the parser in their repo; this makes updating to new versions _much_ less painful. (However, highlight queries tend to be more or less tied to a specific editor, so these are not as easily centralized. Getting together across the wider tree-sitter ecosystem -- Neovim, Emacs, Helix, Zed -- and agreeing on _some_ standards would be great!)

Regarding a list of parsers, one more current (curated since actively used) starting point could be [nvim-treesitter's list of supported parsers](https://github.com/nvim-treesitter/nvim-treesitter#supported-languages), since we do switch parsers when somebody shows a newer one to have significant advantages over an old one.

Rather than making up some "quality metric", adding objective information to the list would already be helpful (and already indicative of activity):
* whether it uses an external `scanner.c` (or even `scanner.cc`)
* whether the repo contains build scripts for easy `make`
* whether the repo contains (reference) queries
* whether the readme itself claims the project to be experimental

**Comment by mattmassicotte** [comment]
@clason I don't want to derail this issue any more, but I'm actually very interested in standardized queries and build systems. Perhaps we could work together on this?

I agree with all four of your points. These seem like the real stumbling blocks for people making use of tree-sitter, after actually locating parser(s) they are interested in using of course.

@ahelwer As for taking ownership of the parser list in the documents, I am absolutely in. I'm just not sure I understand how I should proceed, given there are currently many open PRs about adding parsers. I tried to this recently, but my PR was never merged, and ultimately became stale. I am definitely willing to make another attempt. Is that what you think should be doing?

**Comment by ahelwer** [comment]
Yeah, just rewrite the list(s) however you want, summarize your reasoning in the PR, then ping @ahlinc and he can merge it then close all the other ones.

**Comment by mattmassicotte** [comment]
ok done! https://github.com/tree-sitter/tree-sitter/pull/2037

**Comment by sogaiu** [comment]
@mattmassicotte Thanks for your efforts!

I didn't see a ping for ahlinc in the PR as suggested by ahelwer -- may be it's good to have?  Or may be I misunderstood...

**Comment by mattmassicotte** [comment]
> @mattmassicotte Thanks for your efforts!
> 
> I didn't see a ping for ahlinc in the PR as suggested by ahelwer -- may be it's good to have? Or may be I misunderstood...

As yes thank you! Adding that now.

**Comment by amaanq** [comment]
I'll close this out since this isn't really an issue, but, as most probably know by now, various improvements have been made to the process of setting up a repository for consumption by downstream users (improved bindings files & template for creating a grammar), and we've cleaned up the list of parsers on the docs to only list those under the tree-sitter org, with a reference a list of known parsers on the wiki page. You can't really accurately determine the quality of a grammar automatically, but there are always good indicators as previously mentioned in the thread to figure that out, and if you *are* wondering about the quality of a grammar, it's probably best to open a discussion or issue in the grammar repo itself, and not here since there's not anything upstream can do.
